File: 004/haskell/004.hs
-- Project Euler Problem # 4
-- Largest Palindrome Product
-- Answer:
--
-- Created by Thomas Nelson <tn90ca@gmail.com>
-- Created on 2015-01-22

module Main where

isPalindrome n = show n == (reverse.show $ n)

palindromes min max =
    [a*b | a <- [min..max], b <- [1..a], isPalindrome (a*b)]
    
palindromes2 min max =
    [a*b | a <- [min..max], b <- [min..max], isPalindrome (a*b)]

l3 = maximum $ palindromes 100 999
l4 = maximum $ palindromes2 100 999

main = do
    print (l5)

File: 004/c++/004.cpp
// Project Euler Problem # 4
// Largest Palindrome Product
// Answer: 906609
//
// Created by Thomas Nelson <tn90ca@gmail.com>
// Created on 2015-05-14

#include <iostream>
#include <math.h>

bool isPalindrom (long n) {
    long n1 = n;
    long n2 = 0;

    while (n1 > 0) {
        int a = n1 % 10;
        n1 /= 10;
        n2 *= 10;
        n2 += a;
    }

    if (n == n2) {
        return true;
    } else {
        return false;
    }
}

int main() {
    long best = 0;

    for (int i=100; i<=999; i++) {
        for (int j=100; j<=999; j++) {
            if (isPalindrom(i*j)) {
                if ((i*j) > best) {
                    best = i*j;
                }
            }
        }
    }


    std::cout << "Largest Palindrome Product = " << best << '\n';

    return 0;
}